src/operator/mxnet_op.h

```diff
@@ -448,6 +449,23 @@ struct Kernel<OP, cpu> {
 #endif
   }
 
+  /*!
+   * \brief Launch a generic CPU kernel with native random generator.
+   * \tparam rnd CPU random generator
+   * \tparam Args Varargs type to eventually pass to the OP::Map() functoion
+   * \param N Number of iterations
+   * \param args Varargs to eventually pass to the OP::Map() functoion
+   */
+  template<typename GType, typename ...Args>
+  inline static void LaunchNativeRandomGenerator(mshadow::Stream<cpu> *,
+                                                 common::random::RandGenerator<cpu, GType> *rnd,
+                                                 const int N, Args... args) {
+    // do not use openmp since it does not guarantee the output order.
```

Comment:
Not clear to me what it means that "does not guarantee the output order". I really think we should support openmp here as sampling on CPU is equivalent important to sampling on CPU and we should  not leave a potential speedup of 4-8 on the table. Sampling on CPU is really slow anyway. 
Wouldn't it be natural to use the exact same design pattern as for the GPU case, i.e. a set of preallocated samplers as a GlobalSampler and then assign them to the different threads?

Reply:
I mean, assume the actual sequence of `std:: mt19937` is `0.1, 0.2, 0.3, 0.4`, when it is generated for `arr[4]` with openmp this could become `arr = {0.2, 0.1, 0.4, 0.3}`.

since `std:: mt19937` is thread-safe, I didn't preallocate it for multi-threads. But you're right, the same design could be adopted.

Reply:
yes I think we should use openmp.

Reply:
Adopting the same design would solve this ordering issue. And we don't need to pre-allocate thousands of CPU-samplers, 256 would certainly be enough.